HW2/P5/physics.pyx

```diff
@@ -0,0 +1,156 @@
+            if xIdx < grid_size and yIdx < grid_size and Grid[xIdx,yIdx] < count:
+                acquire(&(locks[Grid[xIdx,yIdx]]))
+                XY2 = &(XY[Grid[xIdx, yIdx], 0])
+                V2 = &(V[Grid[yIdx, xIdx], 0])
+                if overlapping(XY1, XY2, R):
+                    # SUBPROBLEM 4: Add locking
+                    if not moving_apart(XY1, V1, XY2, V2):
+                        collide(XY1, V1, XY2, V2)
+
+                    # give a slight impulse to help separate them
+                    for dim in range(2):
+                        V2[dim] += eps * (XY2[dim] - XY1[dim])
+                release(&(locks[Grid[xIdx,yIdx]]))
+    release(&(locks[i]))
+
+cpdef update(FLOAT[:, ::1] XY,
+             FLOAT[:, ::1] V,
+             UINT[:, ::1] Grid,
+             float R,
+             float grid_spacing,
+             uintptr_t locks_ptr,
+             float t):
+    cdef:
+        int count = XY.shape[0]
+        int i, j, dim
+        FLOAT *XY1, *XY2, *V1, *V2
+        unsigned int x_grid, y_grid
+        # SUBPROBLEM 4: uncomment this code.
+        omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
+
+    assert XY.shape[0] == V.shape[0]
+    assert XY.shape[1] == V.shape[1] == 2
+
+    with nogil:
+        # bounce off of walls
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+        # scheduling.
+        for i in range(count):
+            for dim in range(2):
+                if (((XY[i, dim] < R) and (V[i, dim] < 0)) or
+                    ((XY[i, dim] > 1.0 - R) and (V[i, dim] > 0))):
+                    V[i, dim] *= -1
+
+        # bounce off of each other
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+        # scheduling.
+        for i in range(count):
+            sub_update(XY, V, R, i, count, Grid, grid_spacing, locks)
+
+        # update positions
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
+        #    scheduling).
+        # SUBPROBLEM 2: update the grid values.
+        for i in range(count):
+            # Check to see if ball is within grid
+            if XY[i,0] >= 0 and XY[i,0] <= 1 and XY[i,1] >= 0 and XY[i,1] <= 1:
+                x_grid = <unsigned int>(XY[i,0] / grid_spacing)
+                y_grid = <unsigned int>(XY[i,1] / grid_spacing)
+                Grid[x_grid, y_grid] = -1
+            for dim in range(2):
+                XY[i, dim] += V[i, dim] * t
+            if (XY[i,0]>=0 and XY[i,0]<=1) and (XY[i,1]>=0 and XY[i,1]<=1): #Check to see if ball is within grid
+                x_grid = <unsigned int>(XY[i,0] / grid_spacing)
+                y_grid = <unsigned int>(XY[i,1] / grid_spacing)
+                Grid[x_grid, y_grid] = i
+
+
+def preallocate_locks(num_locks):
+    cdef omp_lock_t *locks = get_N_locks(num_locks)
+    assert 0 != <uintptr_t> <void *> locks, "could not allocate locks"
+    return <uintptr_t> <void *> locks
```

Comment:
I used almost the same methods for all the problems- probably because of how the questions are structured. I agree the Spatial Decomposition contributes to the most speedup, and your spatial re-ordering achieves one of the best speedups that I have seen!